// @supabase/supabase-auth-helpers/nextjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SupabaseClient = exports.supabaseClient = exports.supabaseServerClient = exports.withApiAuth = exports.withPageAuth = exports.withAuthRequired = exports.getUser = void 0;
var tslib_1 = require("tslib");
// Methods
tslib_1.__exportStar(require("./handlers"), exports);
var getUser_1 = require("./utils/getUser");
Object.defineProperty(exports, "getUser", { enumerable: true, get: function () { return tslib_1.__importDefault(getUser_1).default; } });
var withAuthRequired_1 = require("./utils/withAuthRequired");
Object.defineProperty(exports, "withAuthRequired", { enumerable: true, get: function () { return tslib_1.__importDefault(withAuthRequired_1).default; } });
var withPageAuth_1 = require("./utils/withPageAuth");
Object.defineProperty(exports, "withPageAuth", { enumerable: true, get: function () { return tslib_1.__importDefault(withPageAuth_1).default; } });
var withApiAuth_1 = require("./utils/withApiAuth");
Object.defineProperty(exports, "withApiAuth", { enumerable: true, get: function () { return tslib_1.__importDefault(withApiAuth_1).default; } });
var supabaseServerClient_1 = require("./utils/supabaseServerClient");
Object.defineProperty(exports, "supabaseServerClient", { enumerable: true, get: function () { return tslib_1.__importDefault(supabaseServerClient_1).default; } });
var initSupabase_1 = require("./utils/initSupabase");
Object.defineProperty(exports, "supabaseClient", { enumerable: true, get: function () { return initSupabase_1.supabaseClient; } });
Object.defineProperty(exports, "SupabaseClient", { enumerable: true, get: function () { return initSupabase_1.SupabaseClient; } });
